api/cluster.go

```diff
@@ -312,103 +312,136 @@ func (s *Server) peerQuery(ctx context.Context, data cluster.Traceable, name, pa
 		if err != nil {
-			cancel()
-			log.Error(4, "HTTP Render error querying %s%s: %q", peer.GetName(), path, err)
+			log.Error(3, "HTTP peerQuery unable to get peers, %s", err)
+			errorChan <- err
+			return
 		}
-		responses <- struct {
+		log.Debug("HTTP %s across %d instances", name, len(peerGroups)-1)
+
+		reqCtx, cancel := context.WithCancel(ctx)
+		defer cancel()
+
+		originalPeers := make(map[string]struct{}, len(peerGroups))
+		receivedResponses := make(map[int32]struct{}, len(peerGroups))
+
+		responses := make(chan struct {
 			shardGroup int32
 			data       PeerResponse
 			err        error
-		}{shardGroup, PeerResponse{peer, buf}, err}
-	}
-
-	for group, peers := range peerGroups {
-		peer := peers[0]
-		originalPeers[peer.GetName()] = struct{}{}
-		go askPeer(group, peer)
-	}
+		}, 1)
 
-	result := make(map[string]PeerResponse)
-
-	var ticker *time.Ticker
-	var tickChan <-chan time.Time
-	if speculationThreshold != 1 {
-		ticker = time.NewTicker(5 * time.Millisecond)
-		tickChan = ticker.C
-		defer ticker.Stop()
-	}
+		askPeer := func(shardGroup int32, peer cluster.Node) {
+			log.Debug("HTTP Render querying %s%s", peer.GetName(), path)
+			buf, err := peer.Post(reqCtx, name, path, data)
 
-	for len(receivedResponses) < len(peerGroups) {
-		select {
-		case resp := <-responses:
-			if _, ok := receivedResponses[resp.shardGroup]; ok {
-				// already received this response (possibly speculatively)
-				continue
+			select {
+			case <-ctx.Done():
+				return
+			default:
+				// Not canceled, continue
 			}
 
-			if resp.err != nil {
-				return nil, err
+			if err != nil {
+				cancel()
+				log.Error(4, "HTTP Render error querying %s%s: %q", peer.GetName(), path, err)
 			}
+			responses <- struct {
+				shardGroup int32
+				data       PeerResponse
+				err        error
+			}{shardGroup, PeerResponse{peer, buf}, err}
+		}
 
-			result[resp.data.peer.GetName()] = resp.data
-			receivedResponses[resp.shardGroup] = struct{}{}
-			delete(originalPeers, resp.data.peer.GetName())
-
-		case <-tickChan:
-			// Check if it's time to speculate!
-			percentReceived := float64(len(receivedResponses)) / float64(len(peerGroups))
-			if percentReceived >= speculationThreshold {
-				// kick off speculative queries to other members now
-				ticker.Stop()
-				speculativeAttempts.Inc()
-				for shardGroup, peers := range peerGroups {
-					if _, ok := receivedResponses[shardGroup]; ok {
-						continue
-					}
-					eligiblePeers := peers[1:]
-					for _, peer := range eligiblePeers {
-						speculativeRequests.Inc()
-						go askPeer(shardGroup, peer)
+		for group, peers := range peerGroups {
+			peer := peers[0]
+			originalPeers[peer.GetName()] = struct{}{}
+			go askPeer(group, peer)
+		}
+
+		var ticker *time.Ticker
+		var tickChan <-chan time.Time
+		if speculationThreshold != 1 {
+			ticker = time.NewTicker(5 * time.Millisecond)
+			tickChan = ticker.C
+			defer ticker.Stop()
+		}
+
+		for len(receivedResponses) < len(peerGroups) {
```

Comment:
why `for` here?

Reply:
To loop! `for` is go's `while`

Reply:
silly me.  when  I see `for` I expect some sort of range or update  at the declaration point.  for some reason I read this like it was a simple `if`.